File: electron/db/DataManager.ts
/**
 * DataManager：数据管理操作
 * - 一键瘦身：清理已删除 segments + 过期截图 + 孤立数据
 * - 一键清空当天数据
 * - 一键清空全部数据
 * 全部使用参数化 SQL，返回清理统计。
 */
import { app } from 'electron'
import fs from 'node:fs'
import path from 'node:path'
import { getDatabase } from './database'
import { SettingsStore } from './SettingsStore'

export interface CleanupStats {
  deletedSegments: number
  deletedEpisodes: number
  deletedScreenshots: number
  orphanWikiSources: number
}

export interface ClearResult {
  segments: number
  episodes: number
}

/** 截图存储目录（userData/screenshots） */
function getScreenshotsDir(): string {
  return path.join(app.getPath('userData'), 'screenshots')
}

/** 列出截图目录下所有文件 */
function listScreenshotFiles(): string[] {
  const dir = getScreenshotsDir()
  try {
    if (!fs.existsSync(dir)) return []
    return fs.readdirSync(dir).map((f) => path.join(dir, f))
  } catch {
    return []
  }
}

/** 删除指定截图文件（若存在） */
function deleteScreenshotFile(filePath: string): boolean {
  try {
    if (filePath && fs.existsSync(filePath)) {
      fs.unlinkSync(filePath)
      return true
    }
  } catch {
    /* 静默处理 */
  }
  return false
}

export const DataManager = {
  /**
   * 一键瘦身：
   * 1. 物理删除已软删除的 segments 及其截图
   * 2. 删除过期截图（超过保留天数的）
   * 3. 删除孤立 episodes（segmentIds 全部不存在的）
   * 4. 清理 Wiki 中失效的 sources 引用
   */
  cleanup(): CleanupStats {
    const db = getDatabase()
    const stats: CleanupStats = {
      deletedSegments: 0,
      deletedEpisodes: 0,
      deletedScreenshots: 0,
      orphanWikiSources: 0
    }

    // 1. 物理删除已软删除的 segments 及其截图
    const softDeletedRows = db
      .prepare('SELECT screenshot_path FROM segments WHERE is_deleted = 1 AND screenshot_path != ?')
      .all('') as Array<{ screenshot_path: string }>
    for (const row of softDeletedRows) {
      if (deleteScreenshotFile(row.screenshot_path)) {
        stats.deletedScreenshots++
      }
    }
    const delSegResult = db.prepare('DELETE FROM segments WHERE is_deleted = 1').run()
    stats.deletedSegments = delSegResult.changes

    // 2. 删除过期截图（超过保留天数）
    const settings = SettingsStore.get()
    if (settings.saveScreenshots && settings.screenshotRetentionDays > 0) {
      const retentionMs = settings.screenshotRetentionDays * 24 * 60 * 60 * 1000
      const now = Date.now()
      const files = listScreenshotFiles()
      for (const f of files) {
        try {
          const stat = fs.statSync(f)
          if (now - stat.mtimeMs > retentionMs) {
            fs.unlinkSync(f)
            stats.deletedScreenshots++
          }
        } catch {
          /* 跳过无法访问的文件 */
        }
      }
    } else if (!settings.saveScreenshots) {
      // 不保存截图模式：清空整个截图目录
      const files = listScreenshotFiles()
      for (const f of files) {
        if (deleteScreenshotFile(f)) {
          stats.deletedScreenshots++
        }
      }
    }

    // 3. 删除孤立 episodes（segmentIds 全部不存在于 segments 表）
    const episodes = db
      .prepare('SELECT id, segment_ids FROM episodes')
      .all() as Array<{ id: string; segment_ids: string }>
    for (const ep of episodes) {
      let segIds: string[] = []
      try {
        const parsed = JSON.parse(ep.segment_ids)
        if (Array.isArray(parsed)) segIds = parsed as string[]
      } catch {
        /* 解析失败视为孤立 */
      }
      if (segIds.length === 0) continue
      // 检查是否所有 segmentId 都不存在
      const placeholders = segIds.map(() => '?').join(',')
      const existing = db
        .prepare(`SELECT COUNT(*) as cnt FROM segments WHERE id IN (${placeholders})`)
        .get(...segIds) as { cnt: number }
      if (existing.cnt === 0) {
        db.prepare('DELETE FROM episodes WHERE id = ?').run(ep.id)
        stats.deletedEpisodes++
      }
    }

    // 4. 清理 Wiki 中失效的 sources 引用（sources 中的 episode/segment id 已不存在）
    const wikiRows = db
      .prepare('SELECT id, sources FROM wiki_pages')
      .all() as Array<{ id: string; sources: string }>
    for (const row of wikiRows) {
      let srcIds: string[] = []
      try {
        const parsed = JSON.parse(row.sources)
        if (Array.isArray(parsed)) srcIds = parsed as string[]
      } catch {
        /* 解析失败跳过 */
      }
      if (srcIds.length === 0) continue
      const placeholders = srcIds.map(() => '?').join(',')
      const epExisting = db
        .prepare(`SELECT COUNT(*) as cnt FROM episodes WHERE id IN (${placeholders})`)
        .get(...srcIds) as { cnt: number }
      const segExisting = db
        .prepare(`SELECT COUNT(*) as cnt FROM segments WHERE id IN (${placeholders})`)
        .get(...srcIds) as { cnt: number }
      const validCount = epExisting.cnt + segExisting.cnt
      if (validCount < srcIds.length) {
        // 保留仍存在的引用
        const validIds = srcIds.filter((sid) => {
          const ep = db.prepare('SELECT 1 FROM episodes WHERE id = ?').get(sid)
          const seg = db.prepare('SELECT 1 FROM segments WHERE id = ?').get(sid)
          return ep !== undefined || seg !== undefined
        })
        db.prepare('UPDATE wiki_pages SET sources = ? WHERE id = ?').run(
          JSON.stringify(validIds),
          row.id
        )
        stats.orphanWikiSources += srcIds.length - validIds.length
      }
    }

    return stats
  },

  /** 一键清空指定日期的数据（segments + episodes） */
  clearDay(date: string): ClearResult {
    const db = getDatabase()
    // 先收集要删除的截图路径
    const segRows = db
      .prepare('SELECT screenshot_path FROM segments WHERE date = ? AND screenshot_path != ?')
      .all(date, '') as Array<{ screenshot_path: string }>
    for (const row of segRows) {
      deleteScreenshotFile(row.screenshot_path)
    }
    const segResult = db.prepare('DELETE FROM segments WHERE date = ?').run(date)
    const epResult = db.prepare('DELETE FROM episodes WHERE date = ?').run(date)
    return { segments: segResult.changes, episodes: epResult.changes }
  },

  /** 一键清空全部数据（segments + episodes + wiki_pages + reports，保留 privacy_rules） */
  clearAll(): ClearResult & { wikiPages: number; reports: number } {
    const db = getDatabase()
    // 删除所有截图
    const files = listScreenshotFiles()
    for (const f of files) {
      deleteScreenshotFile(f)
    }
    const segResult = db.prepare('DELETE FROM segments').run()
    const epResult = db.prepare('DELETE FROM episodes').run()
    const wikiResult = db.prepare('DELETE FROM wiki_pages').run()
    const reportResult = db.prepare('DELETE FROM reports').run()
    return {
      segments: segResult.changes,
      episodes: epResult.changes,
      wikiPages: wikiResult.changes,
      reports: reportResult.changes
    }
  },

  /** 获取数据统计（供设置页展示） */
  getStats(): {
    segmentCount: number
    episodeCount: number
    wikiCount: number
    reportCount: number
    screenshotCount: number
    dbSizeBytes: number
  } {
    const db = getDatabase()
    const segCount = (db.prepare('SELECT COUNT(*) as c FROM segments').get() as { c: number }).c
    const epCount = (db.prepare('SELECT COUNT(*) as c FROM episodes').get() as { c: number }).c
    const wikiCount = (db.prepare('SELECT COUNT(*) as c FROM wiki_pages').get() as { c: number }).c
    const reportCount = (db.prepare('SELECT COUNT(*) as c FROM reports').get() as { c: number }).c
    const screenshotFiles = listScreenshotFiles()
    let dbSizeBytes = 0
    try {
      const dbPath = path.join(app.getPath('userData'), 'workmemory.db')
      if (fs.existsSync(dbPath)) {
        dbSizeBytes = fs.statSync(dbPath).size
      }
    } catch {
      /* 静默 */
    }
    return {
      segmentCount: segCount,
      episodeCount: epCount,
      wikiCount,
      reportCount,
      screenshotCount: screenshotFiles.length,
      dbSizeBytes
    }
  }
}

File: src-tauri/src/repositories/data_manager.rs
//! DataManager：数据管理操作（对应 electron/db/DataManager.ts）
//!
//! - 一键瘦身：清理已删除 segments + 过期截图 + 孤立数据
//! - 一键清空当天数据
//! - 一键清空全部数据

use std::fs;
use std::path::PathBuf;

use rusqlite::params;

use crate::db::database::get_database;
use crate::repositories::settings_store::{SettingsStore, APP_DATA_DIR};

#[derive(Debug, Clone, serde::Serialize)]
pub struct CleanupStats {
    pub deleted_segments: usize,
    pub deleted_episodes: usize,
    pub deleted_screenshots: usize,
    pub orphan_wiki_sources: usize,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct ClearResult {
    pub segments: usize,
    pub episodes: usize,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct ClearAllResult {
    pub segments: usize,
    pub episodes: usize,
    pub wiki_pages: usize,
    pub reports: usize,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct DataStats {
    pub segment_count: i64,
    pub episode_count: i64,
    pub wiki_count: i64,
    pub report_count: i64,
    pub screenshot_count: usize,
    pub db_size_bytes: u64,
}

/// 截图存储目录（userData/screenshots）
fn get_screenshots_dir() -> PathBuf {
    let dir = APP_DATA_DIR.lock().unwrap();
    match dir.as_ref() {
        Some(p) => p.join("screenshots"),
        None => PathBuf::from("screenshots"),
    }
}

/// 列出截图目录下所有文件
fn list_screenshot_files() -> Vec<PathBuf> {
    let dir = get_screenshots_dir();
    match fs::read_dir(&dir) {
        Ok(entries) => entries.filter_map(|e| e.ok().map(|e| e.path())).collect(),
        Err(_) => Vec::new(),
    }
}

/// 删除指定截图文件
fn delete_screenshot_file(file_path: &PathBuf) -> bool {
    if file_path.exists() {
        match fs::remove_file(file_path) {
            Ok(_) => return true,
            Err(_) => return false,
        }
    }
    false
}

pub struct DataManager;

impl DataManager {
    /// 一键瘦身
    pub fn cleanup() -> anyhow::Result<CleanupStats> {
        let conn = get_database()?;
        let mut stats = CleanupStats {
            deleted_segments: 0,
            deleted_episodes: 0,
            deleted_screenshots: 0,
            orphan_wiki_sources: 0,
        };

        // 1. 物理删除已软删除的 segments 及其截图
        {
            let mut stmt = conn.prepare(
                "SELECT screenshot_path FROM segments WHERE is_deleted = 1 AND screenshot_path != ''",
            )?;
            let rows: Vec<String> = stmt
                .query_map([], |row| row.get(0))?
                .filter_map(|r| r.ok())
                .collect();
            drop(stmt);
            for path in &rows {
                let p = PathBuf::from(path);
                if delete_screenshot_file(&p) {
                    stats.deleted_screenshots += 1;
                }
            }
        }
        let deleted = conn.execute("DELETE FROM segments WHERE is_deleted = 1", [])?;
        stats.deleted_segments = deleted;

        // 2. 删除过期截图
        let settings = SettingsStore::get();
        if settings.save_screenshots && settings.screenshot_retention_days > 0 {
            let retention_ms: u128 = (settings.screenshot_retention_days as u64 * 24 * 60 * 60 * 1000) as u128;
            let now: u128 = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_millis();
            let files = list_screenshot_files();
            for f in &files {
                if let Ok(metadata) = fs::metadata(f) {
                    if let Ok(modified) = metadata.modified() {
                        if let Ok(duration) = modified.duration_since(std::time::UNIX_EPOCH) {
                            if now - duration.as_millis() > retention_ms {
                                if delete_screenshot_file(f) {
                                    stats.deleted_screenshots += 1;
                                }
                            }
                        }
                    }
                }
            }
        } else if !settings.save_screenshots {
            // 不保存截图模式：清空整个截图目录
            let files = list_screenshot_files();
            for f in &files {
                if delete_screenshot_file(f) {
                    stats.deleted_screenshots += 1;
                }
            }
        }

        // 3. 删除孤立 episodes（segmentIds 全部不存在）
        {
            let mut stmt = conn.prepare("SELECT id, segment_ids FROM episodes")?;
            let rows: Vec<(String, String)> = stmt
                .query_map([], |row| Ok((row.get(0)?, row.get(1)?)))?
                .filter_map(|r| r.ok())
                .collect();
            drop(stmt);
            for (ep_id, segment_ids_str) in &rows {
                let seg_ids: Vec<String> = serde_json::from_str(segment_ids_str).unwrap_or_default();
                if seg_ids.is_empty() {
                    continue;
                }
                // 检查是否所有 segmentId 都不存在
                let placeholders: Vec<String> =
                    (0..seg_ids.len()).map(|i| format!("?{}", i + 1)).collect();
                let sql = format!(
                    "SELECT COUNT(*) as cnt FROM segments WHERE id IN ({})",
                    placeholders.join(",")
                );
                let params: Vec<&dyn rusqlite::ToSql> = seg_ids
                    .iter()
                    .map(|id| id as &dyn rusqlite::ToSql)
                    .collect();
                let count: i64 = conn
                    .query_row(&sql, params.as_slice(), |row| row.get(0))
                    .unwrap_or(0);
                if count == 0 {
                    let _ = conn.execute("DELETE FROM episodes WHERE id = ?1", params![ep_id]);
                    stats.deleted_episodes += 1;
                }
            }
        }

        // 4. 清理 Wiki 中失效的 sources 引用
        {
            let mut stmt = conn.prepare("SELECT id, sources FROM wiki_pages")?;
            let rows: Vec<(String, String)> = stmt
                .query_map([], |row| Ok((row.get(0)?, row.get(1)?)))?
                .filter_map(|r| r.ok())
                .collect();
            drop(stmt);
            for (wiki_id, sources_str) in &rows {
                let src_ids: Vec<String> = serde_json::from_str(sources_str).unwrap_or_default();
                if src_ids.is_empty() {
                    continue;
                }
                let placeholders: Vec<String> =
                    (0..src_ids.len()).map(|i| format!("?{}", i + 1)).collect();
                let ep_sql = format!(
                    "SELECT COUNT(*) as cnt FROM episodes WHERE id IN ({})",
                    placeholders.join(",")
                );
                let seg_sql = format!(
                    "SELECT COUNT(*) as cnt FROM segments WHERE id IN ({})",
                    placeholders.join(",")
                );
                let params: Vec<&dyn rusqlite::ToSql> = src_ids
                    .iter()
                    .map(|id| id as &dyn rusqlite::ToSql)
                    .collect();
                let ep_count: i64 = conn
                    .query_row(&ep_sql, params.as_slice(), |row| row.get(0))
                    .unwrap_or(0);
                let seg_count: i64 = conn
                    .query_row(&seg_sql, params.as_slice(), |row| row.get(0))
                    .unwrap_or(0);
                let valid_count = ep_count + seg_count;
                if (valid_count as usize) < src_ids.len() {
                    // 保留仍存在的引用
                    let mut valid_ids: Vec<String> = Vec::new();
                    for sid in &src_ids {
                        let ep_exists: i64 = conn
                            .query_row(
                                "SELECT 1 FROM episodes WHERE id = ?1",
                                params![sid],
                                |row| row.get(0),
                            )
                            .unwrap_or(0);
                        let seg_exists: i64 = conn
                            .query_row(
                                "SELECT 1 FROM segments WHERE id = ?1",
                                params![sid],
                                |row| row.get(0),
                            )
                            .unwrap_or(0);
                        if ep_exists > 0 || seg_exists > 0 {
                            valid_ids.push(sid.clone());
                        }
                    }
                    let valid_str = serde_json::to_string(&valid_ids).unwrap_or_else(|_| "[]".to_string());
                    let _ = conn.execute(
                        "UPDATE wiki_pages SET sources = ?1 WHERE id = ?2",
                        params![valid_str, wiki_id],
                    );
                    stats.orphan_wiki_sources += src_ids.len() - valid_ids.len();
                }
            }
        }

        Ok(stats)
    }

    /// 一键清空指定日期的数据
    pub fn clear_day(date: &str) -> anyhow::Result<ClearResult> {
        let conn = get_database()?;
        // 先收集要删除的截图路径
        {
            let mut stmt = conn.prepare(
                "SELECT screenshot_path FROM segments WHERE date = ?1 AND screenshot_path != ''",
            )?;
            let rows: Vec<String> = stmt
                .query_map(params![date], |row| row.get(0))?
                .filter_map(|r| r.ok())
                .collect();
            drop(stmt);
            for path in &rows {
                let p = PathBuf::from(path);
                delete_screenshot_file(&p);
            }
        }
        let seg_count = conn.execute("DELETE FROM segments WHERE date = ?1", params![date])?;
        let ep_count = conn.execute("DELETE FROM episodes WHERE date = ?1", params![date])?;
        Ok(ClearResult {
            segments: seg_count,
            episodes: ep_count,
        })
    }

    /// 一键清空全部数据
    pub fn clear_all() -> anyhow::Result<ClearAllResult> {
        let conn = get_database()?;
        // 删除所有截图
        let files = list_screenshot_files();
        for f in &files {
            delete_screenshot_file(f);
        }
        let seg_count = conn.execute("DELETE FROM segments", [])?;
        let ep_count = conn.execute("DELETE FROM episodes", [])?;
        let wiki_count = conn.execute("DELETE FROM wiki_pages", [])?;
        let report_count = conn.execute("DELETE FROM reports", [])?;
        Ok(ClearAllResult {
            segments: seg_count,
            episodes: ep_count,
            wiki_pages: wiki_count,
            reports: report_count,
        })
    }

    /// 获取数据统计
    pub fn get_stats() -> anyhow::Result<DataStats> {
        let conn = get_database()?;
        let seg_count: i64 =
            conn.query_row("SELECT COUNT(*) FROM segments", [], |row| row.get(0))?;
        let ep_count: i64 =
            conn.query_row("SELECT COUNT(*) FROM episodes", [], |row| row.get(0))?;
        let wiki_count: i64 =
            conn.query_row("SELECT COUNT(*) FROM wiki_pages", [], |row| row.get(0))?;
        let report_count: i64 =
            conn.query_row("SELECT COUNT(*) FROM reports", [], |row| row.get(0))?;
        let screenshot_files = list_screenshot_files();

        let mut db_size_bytes = 0u64;
        let dir = APP_DATA_DIR.lock().unwrap();
        if let Some(p) = dir.as_ref() {
            let db_path: PathBuf = p.join("workmemory.db");
            if db_path.exists() {
                if let Ok(metadata) = fs::metadata(&db_path) {
                    db_size_bytes = metadata.len();
                }
            }
        }

        Ok(DataStats {
            segment_count: seg_count,
            episode_count: ep_count,
            wiki_count,
            report_count,
            screenshot_count: screenshot_files.len(),
            db_size_bytes,
        })
    }
}
